File: oc/AddressBookDemo/AddressBookDemo/controller/AddContactViewController.m
//
//  AddContactViewController.m
//  AddressBookDemo
//
//  Created by dfang on 2018-1-17.
//  Copyright © 2018年 east. All rights reserved.
//

#import "AddContactViewController.h"

@interface AddContactViewController ()
@property (weak, nonatomic) IBOutlet UITextField *nameField;
@property (weak, nonatomic) IBOutlet UITextField *phoneField;
@property (weak, nonatomic) IBOutlet UIButton *saveBtn;

@end

@implementation AddContactViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}

- (IBAction)textChanged {
    self.saveBtn.enabled = (self.nameField.text.length > 0 && self.phoneField.text.length > 0);
}

- (IBAction)saveBtnClicked {
    if ([self.delegate respondsToSelector:@selector(addContactViewControllerDidSaveContact:)]) {
        Contact *contact = [[Contact alloc] init];
        contact.name = self.nameField.text;
        contact.phone = self.phoneField.text;
        [self.delegate addContactViewControllerDidSaveContact:contact];
    }
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [self.view endEditing:true];
}

// UITextfieldDelegate
- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    [textField resignFirstResponder];
    return YES;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

@end

File: swift/AddressBookDemo/AddressBookDemo/AddContactViewController.swift
//
//  AddContactViewController.swift
//  AddressBookDemo
//
//  Created by dfang on 2018-1-18.
//  Copyright © 2018年 east. All rights reserved.
//

import UIKit

@objc protocol AddContactViewControllerDelegate: NSObjectProtocol {
    @objc func addContactViewControllerDidSaveContact(contact:Contact)
}

class AddContactViewController: UIViewController, UITextFieldDelegate {
    
    var delegate:AddContactViewControllerDelegate?
    
    @IBOutlet weak var nameField: UITextField!
    @IBOutlet weak var phoneField: UITextField!
    @IBOutlet weak var saveBtn: UIButton!
    
    override func viewDidLoad() {
        
    }
    
    @IBAction func textChanged() {
        saveBtn.isEnabled = (nameField.text != nil && nameField.text!.count > 0 && phoneField.text != nil && phoneField.text!.count > 0)
    }
    
    @IBAction func saveBtnClicked() {
        if (delegate!.responds(to: #selector(AddContactViewControllerDelegate.addContactViewControllerDidSaveContact(contact:)))) {
            let contact = Contact.init(name: nameField.text, andPhone: phoneField.text)
            delegate?.addContactViewControllerDidSaveContact(contact: contact)
        }
    }
    
    // UITextfieldDelegate
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        textField.resignFirstResponder()
        return true
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        self.view.endEditing(true)
    }
}
